hw/ip/rom_ctrl/util/scramble_image.py

```diff
@@ -17,54 +18,133 @@
 from util.design.secded_gen import load_secded_config
 
 
-class MemCtrlParams:
+class ScramblingMode(Enum):
+    ROM0 = "base-rom"
+    ROM1 = "second-rom"
+    SRAM = "sram"
+
+
+_UDict = Dict[object, object]
+
+
+class MemoryController:
     def __init__(
-            self, ctrl_name: str,
-            memory_type: str,
-            module_type: str,
-            scr_key: str, scr_key_width: int,
-            nonce: str, nonce_width: int
-    ):
-        '''A memory controller parameters constructor
+            self, mem_ctrl: _UDict, ctrl_name: str, memory_type: str, mode: str):
+        '''A memory controller constructor
+        @mem_ctrl The memory controller hjson dictionary.
         @ctrl_name is the memory controller IP block name (e.g. 'rom_ctrl0' for the base ROM)
         @memory_type is the memory type this memory controller run ('rom' or 'ram')
-        @module_type is the HJSON module type for this memory controller (e.g. 'rom_ctrl')
-        @scr_key is the scrambling key entry name in the HJSON file
-        @scr_key_width is the expected scrambling key width
-        @nonce is the nonce entry name in the HJSON file
-        @nonce_width is the expected nonce width
+        @mode is the memory scrambling mode
         '''
+
+        memory_type, key_name, nonce_name = {
+            ScramblingMode.ROM0.value: ["rom", "RndCnstScrKey", "RndCnstScrNonce"],
+            ScramblingMode.ROM1.value: ["rom", "RndCnstScrKey", "RndCnstScrNonce"],
+            ScramblingMode.SRAM.value: ["ram", "RndCnstSramKey", "RndCnstSramNonce"],
+        }[mode]
+
+        assert memory_type is not None
+        assert key_name is not None
+        assert nonce_name is not None
+
+        size_words = MemoryController._get_size_words(mem_ctrl, memory_type)
+        base = MemoryController._get_base(mem_ctrl, memory_type)
+        params = MemoryController._get_params(mem_ctrl)
+        nonce, nonce_width = MemoryController._get_param_cnst(params, nonce_name)
+        scr_key, scr_key_width = MemoryController._get_param_cnst(params, key_name)
+
         self.ctrl_name = ctrl_name
         self.memory_type = memory_type
-        self.module_type = module_type
+        self.base = base
+        self.size_words = size_words
         self.scr_key = scr_key
         self.scr_key_width = scr_key_width
         self.nonce = nonce
         self.nonce_width = nonce_width
 
+    @staticmethod
+    def _get_params(module: _UDict) -> Dict[str, _UDict]:
+        params = module.get('param_list')
+        assert isinstance(params, list)
+
+        named_params = {}  # type: Dict[str, _UDict]
+        for param in params:
+            name = param.get('name')
+            assert isinstance(name, str)
+            assert name not in named_params
+            named_params[name] = param
+
+        return named_params
+
+    @staticmethod
+    def _get_param_cnst(params: Dict[str, _UDict], name: str) -> tuple[int, int]:
```

Comment:
This is causing a CI failure. I think you want `Tuple[int, int]`.

Reply:
Weird, it seemed ok locally. Anyways, this should be fixed now.